File: python/src/onchaindiligence/agent_evidence/trust.py
"""Explicit caller-supplied trust policy and key lifecycle evaluation."""

from __future__ import annotations

import base64
import re
from collections.abc import Iterable, Mapping
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from hashlib import sha256
from types import MappingProxyType
from typing import Any

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from .canonical import format_timestamp, parse_timestamp
from .errors import ParseError, SchemaValidationError, TrustPolicyError
from .models import JsonObject, VerificationState
from .schema import validate_document

_KEY_ID = re.compile(r"^ed25519-[A-Za-z0-9_-]{16}$")


def derive_key_id(public_key: Ed25519PublicKey) -> str:
    """Derive the required key ID from Ed25519 SPKI DER bytes."""

    der = public_key.public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    digest = base64.urlsafe_b64encode(sha256(der).digest()).decode("ascii").rstrip("=")
    return f"ed25519-{digest[:16]}"


def load_public_key_pem(value: str | bytes) -> Ed25519PublicKey:
    """Load an Ed25519 public key, rejecting every other algorithm."""

    raw = value.encode("utf-8") if isinstance(value, str) else value
    try:
        key = serialization.load_pem_public_key(raw)
    except (TypeError, ValueError) as exc:
        raise TrustPolicyError("public_key_pem is not a valid SubjectPublicKeyInfo PEM") from exc
    if not isinstance(key, Ed25519PublicKey):
        raise TrustPolicyError("public_key_pem must contain an Ed25519 public key")
    return key


def create_key_record(
    public_key: Ed25519PublicKey,
    *,
    valid_from: str | datetime | None,
    status: str = "active",
    valid_until: str | datetime | None = None,
    status_changed_at: str | datetime | None = None,
    replacement_key_id: str | None = None,
    compromised_at: str | datetime | None = None,
    status_reason: str | None = None,
) -> JsonObject:
    """Create and validate a portable lifecycle record for a public key."""

    def timestamp(value: str | datetime | None) -> str | None:
        if isinstance(value, datetime):
            return format_timestamp(value)
        return value

    pem = public_key.public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode("ascii")
    changed = status_changed_at
    if changed is None and status == "active":
        changed = valid_from
    record: JsonObject = {
        "key_id": derive_key_id(public_key),
        "algorithm": "ed25519",
        "public_key_pem": pem,
        "status": status,
        "valid_from": timestamp(valid_from),
        "valid_until": timestamp(valid_until),
        "status_changed_at": timestamp(changed),
        "replacement_key_id": replacement_key_id,
        "compromised_at": timestamp(compromised_at),
    }
    if status_reason is not None:
        record["status_reason"] = status_reason
    return AttestationKey.from_record(record).to_record()


def _optional_timestamp(record: Mapping[str, Any], name: str) -> datetime | None:
    value = record.get(name)
    if value is None:
        return None
    if not isinstance(value, str):
        raise TrustPolicyError(f"{name} must be an exact timestamp or null")
    try:
        return parse_timestamp(value)
    except ParseError as exc:
        raise TrustPolicyError(f"invalid key {name}: {exc}") from exc


@dataclass(frozen=True, slots=True)
class AttestationKey:
    """Validated caller-trusted Ed25519 lifecycle record."""

    key_id: str
    public_key: Ed25519PublicKey = field(repr=False, compare=False)
    public_key_pem: str
    status: str
    valid_from: datetime | None
    valid_until: datetime | None
    status_changed_at: datetime | None
    replacement_key_id: str | None = None
    compromised_at: datetime | None = None
    status_reason: str | None = None

    @classmethod
    def from_record(cls, record: Mapping[str, Any]) -> AttestationKey:
        """Validate schema, SPKI identity, timestamps, and lifecycle coherence."""

        material = dict(record)
        try:
            validate_document("attestation-key.schema.json", material)
        except SchemaValidationError as exc:
            raise TrustPolicyError(str(exc)) from exc
        key_id = material.get("key_id")
        pem = material.get("public_key_pem")
        if not isinstance(key_id, str) or not _KEY_ID.fullmatch(key_id):
            raise TrustPolicyError("invalid Ed25519 key_id")
        if not isinstance(pem, str):
            raise TrustPolicyError("public_key_pem must be a string")
        public_key = load_public_key_pem(pem)
        derived = derive_key_id(public_key)
        if derived != key_id:
            raise TrustPolicyError(f"key_id does not match SPKI public key: expected {derived}")

        valid_from = _optional_timestamp(material, "valid_from")
        valid_until = _optional_timestamp(material, "valid_until")
        changed = _optional_timestamp(material, "status_changed_at")
        compromised = _optional_timestamp(material, "compromised_at")
        status = str(material["status"])
        if valid_from is not None and valid_until is not None and valid_until < valid_from:
            raise TrustPolicyError("valid_until precedes valid_from")
        if valid_from is not None and changed is not None and changed < valid_from:
            raise TrustPolicyError("status_changed_at precedes valid_from")
        if valid_from is not None and compromised is not None and compromised < valid_from:
            raise TrustPolicyError("compromised_at precedes valid_from")
        if status in {"revoked", "compromised"} and changed is None:
            raise TrustPolicyError(f"{status} key requires status_changed_at")
        replacement_key_id = material.get("replacement_key_id")
        if replacement_key_id == key_id:
            raise TrustPolicyError("a key cannot name itself as its replacement")

        return cls(
            key_id=key_id,
            public_key=public_key,
            public_key_pem=pem,
            status=status,
            valid_from=valid_from,
            valid_until=valid_until,
            status_changed_at=changed,
            replacement_key_id=replacement_key_id,
            compromised_at=compromised,
            status_reason=material.get("status_reason"),
        )

    def to_record(self) -> JsonObject:
        """Return the portable public record shape."""

        def timestamp(value: datetime | None) -> str | None:
            if value is None:
                return None
            return value.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")

        record: JsonObject = {
            "key_id": self.key_id,
            "algorithm": "ed25519",
            "public_key_pem": self.public_key_pem,
            "status": self.status,
            "valid_from": timestamp(self.valid_from),
            "valid_until": timestamp(self.valid_until),
            "status_changed_at": timestamp(self.status_changed_at),
            "replacement_key_id": self.replacement_key_id,
            "compromised_at": timestamp(self.compromised_at),
        }
        if self.status_reason is not None:
            record["status_reason"] = self.status_reason
        return record


@dataclass(frozen=True, slots=True)
class TrustPolicy:
    """All trust and freshness inputs used by the zero-network verifier."""

    keys: Mapping[str, AttestationKey] = field(default_factory=dict)
    now: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    required_signature_key_ids: frozenset[str] = field(default_factory=frozenset)
    minimum_valid_signatures: int = 1
    max_future_skew: timedelta = timedelta(minutes=5)
    max_bundle_age: timedelta | None = None
    enforce_evidence_expiration: bool = True
    allow_digest_only_evidence: bool = False
    require_verified_anchor: bool = False
    max_file_size: int = 10 * 1024 * 1024
    max_depth: int = 64
    max_string_length: int = 1024 * 1024
    max_array_length: int = 10_000

    def __post_init__(self) -> None:
        if self.now.tzinfo is None:
            raise TrustPolicyError("policy now must be timezone-aware")
        if self.minimum_valid_signatures < 1:
            raise TrustPolicyError("minimum_valid_signatures must be at least one")
        if self.max_future_skew < timedelta(0):
            raise TrustPolicyError("max_future_skew cannot be negative")
        if self.max_bundle_age is not None and self.max_bundle_age < timedelta(0):
            raise TrustPolicyError("max_bundle_age cannot be negative")
        if min(self.max_file_size, self.max_depth, self.max_string_length, self.max_array_length) < 1:
            raise TrustPolicyError("parser limits must be positive")
        copied = dict(self.keys)
        for key_id, key in copied.items():
            if key_id != key.key_id:
                raise TrustPolicyError("trust mapping key does not match key record")
        object.__setattr__(self, "keys", MappingProxyType(copied))
        object.__setattr__(self, "now", self.now.astimezone(timezone.utc))
        missing_required = self.required_signature_key_ids - copied.keys()
        if missing_required:
            raise TrustPolicyError(
                "required signature keys are absent from caller trust: " + ", ".join(sorted(missing_required))
            )

    @classmethod
    def from_key_records(
        cls,
        records: Iterable[Mapping[str, Any]],
        **options: Any,
    ) -> TrustPolicy:
        """Build an explicit trust policy; embedded bundle keys are never consulted."""

        keys: dict[str, AttestationKey] = {}
        for raw in records:
            key = AttestationKey.from_record(raw)
            if key.key_id in keys:
                raise TrustPolicyError(f"duplicate caller-trusted key: {key.key_id}")
            keys[key.key_id] = key
        for key in keys.values():
            if key.replacement_key_id is not None and key.replacement_key_id not in keys:
                raise TrustPolicyError(
                    f"key {key.key_id} references an absent replacement key: {key.replacement_key_id}"
                )
        return cls(keys=keys, **options)


def evaluate_key_lifecycle(
    key: AttestationKey,
    *,
    signed_at: datetime | None,
    policy: TrustPolicy,
) -> tuple[VerificationState, str, str]:
    """Evaluate status/window without treating asserted time as external time proof."""

    if key.status == "revoked":
        return VerificationState.INVALID, "key-revoked", "the caller-trusted key is revoked"
    if key.status == "compromised":
        return VerificationState.INVALID, "key-compromised", "the caller-trusted key is compromised"
    if key.valid_from is None:
        return (
            VerificationState.UNVERIFIABLE,
            "key-valid-from-missing",
            "the caller-trusted key has no defensible activation boundary",
        )
    if signed_at is None:
        if key.status == "retired":
            return (
                VerificationState.UNVERIFIABLE,
                "signature-time-missing",
                "a retired key cannot validate a proof that carries no signed issuance time",
            )
        if policy.now < key.valid_from:
            return (
                VerificationState.INVALID,
                "key-not-yet-valid",
                "the key is not active at policy time",
            )
        return VerificationState.VALID, "key-active", "the caller-trusted key is currently active"
    if signed_at > policy.now + policy.max_future_skew:
        return (
            VerificationState.INVALID,
            "signature-time-future",
            "signed time exceeds allowed clock skew",
        )
    if signed_at < key.valid_from:
        return VerificationState.INVALID, "key-not-yet-valid", "signed time precedes key activation"
    if key.valid_until is not None and signed_at > key.valid_until:
        return (
            VerificationState.INVALID,
            "key-expired",
            "signed time follows the key validity interval",
        )
    return (
        VerificationState.VALID,
        "key-window-valid",
        "signed time is inside the key validity interval",
    )

File: packages/agent-evidence/src/trust.ts
import {
  createHash,
  createPublicKey,
  KeyObject,
} from 'node:crypto'
import { formatTimestamp, parseTimestamp } from './canonical.js'
import { SchemaValidationError, TrustPolicyError } from './errors.js'
import { validateDocument } from './schema.js'
import type { AttestationKeyRecord, KeyInput, VerificationState } from './types.js'

const KEY_ID = /^ed25519-[A-Za-z0-9_-]{16}$/

function asBuffer(value: Uint8Array): Buffer {
  return Buffer.from(value.buffer, value.byteOffset, value.byteLength)
}

export function loadPublicKey(value: KeyInput): KeyObject {
  try {
    const key = value instanceof KeyObject
      ? value
      : value instanceof Uint8Array && !Buffer.isBuffer(value)
      ? createPublicKey(asBuffer(value))
      : createPublicKey(value as string | Buffer)
    if (key.type !== 'public' || key.asymmetricKeyType !== 'ed25519') {
      throw new TrustPolicyError('public key must be Ed25519')
    }
    return key
  } catch (error) {
    if (error instanceof TrustPolicyError) throw error
    throw new TrustPolicyError('public key is not a valid Ed25519 SubjectPublicKeyInfo key', { cause: error })
  }
}

export function deriveKeyId(publicKey: KeyInput): string {
  const key = loadPublicKey(publicKey)
  const der = key.export({ format: 'der', type: 'spki' })
  return `ed25519-${createHash('sha256').update(der).digest('base64url').slice(0, 16)}`
}

function optionalTimestamp(record: AttestationKeyRecord, name: keyof AttestationKeyRecord): Date | null {
  const value = record[name]
  if (value === null || value === undefined) return null
  if (typeof value !== 'string') throw new TrustPolicyError(`${String(name)} must be an exact timestamp or null`)
  try {
    return parseTimestamp(value)
  } catch (error) {
    throw new TrustPolicyError(`invalid key ${String(name)}: ${error instanceof Error ? error.message : String(error)}`, {
      cause: error,
    })
  }
}

export class AttestationKey {
  readonly keyId: string
  readonly publicKey: KeyObject
  readonly publicKeyPem: string
  readonly status: AttestationKeyRecord['status']
  readonly replacementKeyId: string | null
  readonly statusReason?: string
  readonly #validFromMs: number | null
  readonly #validUntilMs: number | null
  readonly #statusChangedAtMs: number | null
  readonly #compromisedAtMs: number | null

  private constructor(record: AttestationKeyRecord, publicKey: KeyObject) {
    this.keyId = record.key_id
    this.publicKey = publicKey
    this.publicKeyPem = record.public_key_pem
    this.status = record.status
    this.#validFromMs = optionalTimestamp(record, 'valid_from')?.getTime() ?? null
    this.#validUntilMs = optionalTimestamp(record, 'valid_until')?.getTime() ?? null
    this.#statusChangedAtMs = optionalTimestamp(record, 'status_changed_at')?.getTime() ?? null
    this.replacementKeyId = record.replacement_key_id
    this.#compromisedAtMs = optionalTimestamp(record, 'compromised_at')?.getTime() ?? null
    if (record.status_reason !== undefined) this.statusReason = record.status_reason
  }

  get validFrom(): Date | null { return this.#validFromMs === null ? null : new Date(this.#validFromMs) }
  get validUntil(): Date | null { return this.#validUntilMs === null ? null : new Date(this.#validUntilMs) }
  get statusChangedAt(): Date | null {
    return this.#statusChangedAtMs === null ? null : new Date(this.#statusChangedAtMs)
  }
  get compromisedAt(): Date | null {
    return this.#compromisedAtMs === null ? null : new Date(this.#compromisedAtMs)
  }

  static fromRecord(input: AttestationKeyRecord | Record<string, unknown>): AttestationKey {
    const record = structuredClone(input) as AttestationKeyRecord
    try {
      validateDocument('attestation-key.schema.json', record)
    } catch (error) {
      if (error instanceof SchemaValidationError) throw new TrustPolicyError(error.message, { cause: error })
      throw error
    }
    if (!KEY_ID.test(record.key_id)) throw new TrustPolicyError('invalid Ed25519 key_id')
    const publicKey = loadPublicKey(record.public_key_pem)
    const derived = deriveKeyId(publicKey)
    if (derived !== record.key_id) {
      throw new TrustPolicyError(`key_id does not match SPKI public key: expected ${derived}`)
    }
    const key = new AttestationKey(record, publicKey)
    if (key.validFrom && key.validUntil && key.validUntil.getTime() < key.validFrom.getTime()) {
      throw new TrustPolicyError('valid_until precedes valid_from')
    }
    if (key.validFrom && key.statusChangedAt && key.statusChangedAt.getTime() < key.validFrom.getTime()) {
      throw new TrustPolicyError('status_changed_at precedes valid_from')
    }
    if (key.validFrom && key.compromisedAt && key.compromisedAt.getTime() < key.validFrom.getTime()) {
      throw new TrustPolicyError('compromised_at precedes valid_from')
    }
    if ((key.status === 'revoked' || key.status === 'compromised') && !key.statusChangedAt) {
      throw new TrustPolicyError(`${key.status} key requires status_changed_at`)
    }
    if (key.replacementKeyId === key.keyId) throw new TrustPolicyError('a key cannot name itself as its replacement')
    return key
  }

  toRecord(): AttestationKeyRecord {
    const record: AttestationKeyRecord = {
      key_id: this.keyId,
      algorithm: 'ed25519',
      public_key_pem: this.publicKeyPem,
      status: this.status,
      valid_from: this.validFrom?.toISOString() ?? null,
      valid_until: this.validUntil?.toISOString() ?? null,
      status_changed_at: this.statusChangedAt?.toISOString() ?? null,
      replacement_key_id: this.replacementKeyId,
      compromised_at: this.compromisedAt?.toISOString() ?? null,
    }
    if (this.statusReason !== undefined) record.status_reason = this.statusReason
    return record
  }
}

export interface CreateKeyRecordOptions {
  validFrom: string | Date | null
  status?: AttestationKeyRecord['status']
  validUntil?: string | Date | null
  statusChangedAt?: string | Date | null
  replacementKeyId?: string | null
  compromisedAt?: string | Date | null
  statusReason?: string
}

function timestamp(value: string | Date | null | undefined): string | null {
  if (value instanceof Date) return formatTimestamp(value)
  return value ?? null
}

export function createKeyRecord(publicKeyInput: KeyInput, options: CreateKeyRecordOptions): AttestationKeyRecord {
  const publicKey = loadPublicKey(publicKeyInput)
  const status = options.status ?? 'active'
  const validFrom = timestamp(options.validFrom)
  const changed = options.statusChangedAt === undefined && status === 'active'
    ? validFrom
    : timestamp(options.statusChangedAt)
  const record: AttestationKeyRecord = {
    key_id: deriveKeyId(publicKey),
    algorithm: 'ed25519',
    public_key_pem: publicKey.export({ format: 'pem', type: 'spki' }).toString(),
    status,
    valid_from: validFrom,
    valid_until: timestamp(options.validUntil),
    status_changed_at: changed,
    replacement_key_id: options.replacementKeyId ?? null,
    compromised_at: timestamp(options.compromisedAt),
  }
  if (options.statusReason !== undefined) record.status_reason = options.statusReason
  return AttestationKey.fromRecord(record).toRecord()
}

export interface TrustPolicyOptions {
  now?: Date
  requiredSignatureKeyIds?: Iterable<string>
  minimumValidSignatures?: number
  maxFutureSkewMs?: number
  maxBundleAgeMs?: number | null
  enforceEvidenceExpiration?: boolean
  allowDigestOnlyEvidence?: boolean
  requireVerifiedAnchor?: boolean
  maxFileSize?: number
  maxDepth?: number
  maxStringLength?: number
  maxArrayLength?: number
}

export class TrustPolicy {
  readonly minimumValidSignatures: number
  readonly maxFutureSkewMs: number
  readonly maxBundleAgeMs: number | null
  readonly enforceEvidenceExpiration: boolean
  readonly allowDigestOnlyEvidence: boolean
  readonly requireVerifiedAnchor: boolean
  readonly maxFileSize: number
  readonly maxDepth: number
  readonly maxStringLength: number
  readonly maxArrayLength: number
  readonly #keys: ReadonlyMap<string, AttestationKey>
  readonly #nowMs: number
  readonly #requiredSignatureKeyIds: ReadonlySet<string>

  private constructor(keys: Map<string, AttestationKey>, options: TrustPolicyOptions) {
    this.#nowMs = options.now?.getTime() ?? Date.now()
    if (Number.isNaN(this.#nowMs)) throw new TrustPolicyError('policy now must be a valid Date')
    this.#requiredSignatureKeyIds = new Set(options.requiredSignatureKeyIds ?? [])
    this.minimumValidSignatures = options.minimumValidSignatures ?? 1
    this.maxFutureSkewMs = options.maxFutureSkewMs ?? 5 * 60 * 1000
    this.maxBundleAgeMs = options.maxBundleAgeMs ?? null
    this.enforceEvidenceExpiration = options.enforceEvidenceExpiration ?? true
    this.allowDigestOnlyEvidence = options.allowDigestOnlyEvidence ?? false
    this.requireVerifiedAnchor = options.requireVerifiedAnchor ?? false
    this.maxFileSize = options.maxFileSize ?? 10 * 1024 * 1024
    this.maxDepth = options.maxDepth ?? 64
    this.maxStringLength = options.maxStringLength ?? 1024 * 1024
    this.maxArrayLength = options.maxArrayLength ?? 10_000
    if (!Number.isInteger(this.minimumValidSignatures) || this.minimumValidSignatures < 1) {
      throw new TrustPolicyError('minimumValidSignatures must be at least one')
    }
    if (this.maxFutureSkewMs < 0) throw new TrustPolicyError('maxFutureSkewMs cannot be negative')
    if (this.maxBundleAgeMs !== null && this.maxBundleAgeMs < 0) {
      throw new TrustPolicyError('maxBundleAgeMs cannot be negative')
    }
    if ([this.maxFileSize, this.maxDepth, this.maxStringLength, this.maxArrayLength].some((value) => value < 1)) {
      throw new TrustPolicyError('parser limits must be positive')
    }
    const missing = [...this.#requiredSignatureKeyIds].filter((keyId) => !keys.has(keyId))
    if (missing.length) {
      throw new TrustPolicyError(`required signature keys are absent from caller trust: ${missing.sort().join(', ')}`)
    }
    this.#keys = new Map(keys)
  }

  static fromKeyRecords(
    records: Iterable<AttestationKeyRecord | Record<string, unknown>>,
    options: TrustPolicyOptions = {},
  ): TrustPolicy {
    const keys = new Map<string, AttestationKey>()
    for (const record of records) {
      const key = AttestationKey.fromRecord(record)
      if (keys.has(key.keyId)) throw new TrustPolicyError(`duplicate caller-trusted key: ${key.keyId}`)
      keys.set(key.keyId, key)
    }
    for (const key of keys.values()) {
      if (key.replacementKeyId && !keys.has(key.replacementKeyId)) {
        throw new TrustPolicyError(
          `key ${key.keyId} references an absent replacement key: ${key.replacementKeyId}`,
        )
      }
    }
    return new TrustPolicy(keys, options)
  }

  key(keyId: string): AttestationKey | undefined {
    return this.#keys.get(keyId)
  }

  get now(): Date { return new Date(this.#nowMs) }

  get requiredSignatureKeyIds(): ReadonlySet<string> {
    return new Set(this.#requiredSignatureKeyIds)
  }
}

export function evaluateKeyLifecycle(
  key: AttestationKey,
  signedAt: Date | null,
  policy: TrustPolicy,
): readonly [VerificationState, string, string] {
  if (key.status === 'revoked') return ['INVALID', 'key-revoked', 'the caller-trusted key is revoked']
  if (key.status === 'compromised') return ['INVALID', 'key-compromised', 'the caller-trusted key is compromised']
  if (!key.validFrom) {
    return ['UNVERIFIABLE', 'key-valid-from-missing', 'the caller-trusted key has no defensible activation boundary']
  }
  if (!signedAt) {
    if (key.status === 'retired') {
      return ['UNVERIFIABLE', 'signature-time-missing', 'a retired key cannot validate a proof that carries no signed issuance time']
    }
    if (policy.now.getTime() < key.validFrom.getTime()) {
      return ['INVALID', 'key-not-yet-valid', 'the key is not active at policy time']
    }
    return ['VALID', 'key-active', 'the caller-trusted key is currently active']
  }
  if (signedAt.getTime() > policy.now.getTime() + policy.maxFutureSkewMs) {
    return ['INVALID', 'signature-time-future', 'signed time exceeds allowed clock skew']
  }
  if (signedAt.getTime() < key.validFrom.getTime()) {
    return ['INVALID', 'key-not-yet-valid', 'signed time precedes key activation']
  }
  if (key.validUntil && signedAt.getTime() > key.validUntil.getTime()) {
    return ['INVALID', 'key-expired', 'signed time follows the key validity interval']
  }
  return ['VALID', 'key-window-valid', 'signed time is inside the key validity interval']
}
